[WIP] Span batch implementation

ImTei opened the pull request:
Everything is in progress
Code structure and naming may be changed

pcw109550 committed:
Add span batch and span batch decoder

pcw109550 committed:
Fix for supporting span batch

pcw109550 committed:
Fix error message

pcw109550 committed:
Add span batch encoder

pcw109550 committed:
Memory optimization

pcw109550 committed:
Remove magic number

pcw109550 committed:
Memory optimization

pcw109550 committed:
Enforce timestamp to be encoded in 8 bytes

pcw109550 committed:
Marshal batch version

pcw109550 committed:
Fix bunch of bugs

pcw109550 committed:
Add test for span batch serialization

pcw109550 committed:
Memory optimization

pcw109550 committed:
Initialize batch merge method

pcw109550 committed:
Tx data encoding for batch merge method

pcw109550 committed:
Refactor tx data encoding

pcw109550 committed:
Add l1OriginNum field to BatchV2Prefix

pcw109550 committed:
Apply span batch spec update

pcw109550 committed:
Fix first bit of OriginBits

pcw109550 committed:
Add validation TODOs when decoding

pcw109550 committed:
Add method for converting BatchV2Tx to TxData

pcw109550 committed:
Add batch split method

pcw109550 committed:
Refactor batch test

pcw109550 committed:
Add RandomBatchV1 method

pcw109550 committed:
Refactor batch test

pcw109550 committed:
Refactor batch test

pcw109550 committed:
Sort BatchV1 by L2 timestamp while merging

pcw109550 committed:
Make txData more realistic

pcw109550 committed:
Fix wrong type conversion

pcw109550 committed:
Decrement l1OriginBlockNumber while BatchV2 splitting

pcw109550 committed:
Memory optimization

pcw109550 committed:
Fix split BatchV2 bug

pcw109550 committed:
Add test for BatchV2 split merge

pcw109550 committed:
Init BatchV2 tx test

pcw109550 committed:
Recover first BatchV1 parent hash

pcw109550 committed:
Test BatchV2 split merge without additional input

pcw109550 committed:
Follow spec definition

pcw109550 committed:
Add invariant test

pcw109550 committed:
Randomize tx types

pcw109550 committed:
Change test name

pcw109550 committed:
Add batchV2 tests for error cases

pcw109550 committed:
Use relTimestamp

pcw109550 committed:
Update test to use relTimestamp

pcw109550 committed:
Separate BatchV2 split validation

pcw109550 committed:
Refactor tests to be DRY

pcw109550 committed:
Fix typo

pcw109550 committed:
Add tests for batchV2 split and merge

pcw109550 committed:
Use uvarint in BatchV2 prefix

pcw109550 committed:
Update comments

ImTei committed:
BatchV2 derivation

ImTei committed:
Do not use BatchWithL1InclusionBlock in BatchReader

ImTei committed:
Span batch hardfork config

Commit:
Merge pull request #5 from testinprod-io/pcw109550/span-batch-datastructure

Span Batch Data Structure

pcw109550 commented:
Please change the base branch to [tip/span-batch](https://github.com/testinprod-io/optimism/tree/tip/span-batch) because initial implementation of span batch(pcw109550/span-batch-datastructure) got merged to it.

ImTei committed:
op-batcher WIP

ImTei committed:
Pass batchType and ParentRef in ChannelConfig

ImTei committed:
fetch safe head when creating channel manager

ImTei committed:
Fix shadow compressor and test case

ImTei committed:
pass safehead in NewChannelManager

ImTei committed:
Run channel manager test cases for each batch types

ImTei committed:
Pass channel manager test cases with span batch

ImTei committed:
Run channel_test

ImTei committed:
Run channel_builder_test

ImTei committed:
Fix ChanneOut

ImTei committed:
Run derivation test cases

ImTei committed:
Run derivation test cases without span batch

pcw109550 committed:
RLP hack to read tx data length from itself

pcw109550 committed:
Remove tx data headers

Commit:
Merge pull request #9 from testinprod-io/pcw109550/span-batch-remove-tx-data-header

Remove Transaction data header from Span Batch Payload

pcw109550 commented:
@ImTei Please backmerge https://github.com/testinprod-io/optimism/tree/tip/span-batch since pcw109550/span-batch-remove-tx-data-header got merged(https://github.com/testinprod-io/optimism/pull/9) to `tip/span-batch` branch.

pcw109550 committed:
Update comments

ImTei committed:
Merge branch 'tip/span-batch' into tei/span-batch-research

ImTei committed:
Remove TxDataHeaders

ImTei committed:
Check ErrMaxFrameIndex

ImTei committed:
Rename batch types

ImTei committed:
Replace BatchData with Batch interface

ImTei committed:
Parameterize batch_queue_test

ImTei committed:
Rename InitBatchData

ImTei committed:
Revert "Check ErrMaxFrameIndex"

This reverts commit 589758a11c51fdb3ef0abcd5d3e0d3124e7775ba.

ImTei committed:
Fix MaxRLPBytesPerChannel to cover span batch

ImTei committed:
Fix batcher test cases

ImTei committed:
ChannelOut refactoring WIP

ImTei committed:
Implement SpanBatchBuilder

ImTei committed:
Revert compressor diff

ImTei committed:
Use readyBuf in ChannelOut

ImTei committed:
Revert shadow_compressor_test

ImTei committed:
Fix channelOut, add comments

ImTei committed:
use timeout context in fetchSyncStatus

ImTei committed:
batch files rerorg

ImTei committed:
Fix ChannelOut, add ChannelBuilder test case

ImTei committed:
Rename readyBuf to reader

ImTei committed:
Add span batch config in e2e test

ImTei committed:
Add span batch env flag for e2e

ImTei committed:
Implement span batch rules

ImTei committed:
Update bq.l1Blocks

ImTei committed:
Span batch validation test

ImTei committed:
Implement new span batch specs

ImTei committed:
Rename MergeSingularBatches to NewSpanBatch

ImTei committed:
Support FeeRecipients

ImTei committed:
Rename new types to be private

ImTei committed:
Fix e2e test

ImTei committed:
New span batch specs

ImTei committed:
Span batch refactoring

ImTei committed:
Set batch type when deriving span batch

ImTei committed:
Replace SingularBatch with spanBatchElement in SpanBatch

ImTei committed:
Update parentCheck in AppendSingularBatch

ImTei committed:
Adding comments

ImTei committed:
Remove FeeRecipients

ImTei committed:
Merge branch 'develop' into tei/span-batch-research

ImTei committed:
Fix after develop merge

ImTei committed:
Adding comments

pcw109550 commented on op-node/rollup/derive/singular_batch.go:
```diff
@@ -0,0 +1,64 @@
+	EpochNum   rollup.Epoch // aka l1 num
+	EpochHash  common.Hash  // block hash
+	Timestamp  uint64
+	// no feeRecipient address input, all fees go to a L2 contract
```
Remove mention of `feeRecipient`.

pcw109550 commented on op-node/rollup/derive/singular_batch.go:
```diff
@@ -0,0 +1,64 @@
+type SingularBatch struct {
+	ParentHash common.Hash  // parent L2 block hash
+	EpochNum   rollup.Epoch // aka l1 num
+	EpochHash  common.Hash  // block hash
```
```suggestion
	EpochHash  common.Hash  // l1 block hash
```

pcw109550 commented on op-node/rollup/derive/singular_batch.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+// CheckOriginHash checks if the epoch hash (L1 origin block hash) matches the given hash, probably L1 block hash from the current canonical L1 chain.
+func (b *SingularBatch) CheckOriginHash(hash common.Hash) bool {
```
Why the name `CheckOriginHash`, not `CheckEpochHash`?

pcw109550 commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,533 @@
+	l1OriginCheck []byte // First 20 bytes of the last block's L1 origin hash
+}
+
+type spanBatchSignature struct {
```
Suggestion to move this struct to `span_batch_txs.go`.

pcw109550 commented:
Can we place all batch related files to batch directory? `op-node/rollup/derive/batch/*`.

pcw109550 commented:
We must remove all the TODO comments: to validate each fields.

pcw109550 commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,533 @@
+}
+
+// GetEpochNum returns epoch number(L1 origin block number) of the first block in the span
+func (b *SpanBatch) GetEpochNum() rollup.Epoch {
```
Are we sure that `GetEpochNum` return first block's l1 origin? According to the spec, https://github.com/testinprod-io/optimism/blob/develop/specs/span-batches.md, 
> l1_origin_num: uvarint number of last l1 origin number. i.e. span_end.l1_origin.number

pcw109550 replied:
I think this is intended.

pcw109550 replied:
@ImTei will research appropriate method name.

pcw109550 commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,533 @@
+func (b *SpanBatch) AppendSingularBatch(singularBatch *SingularBatch) {
+	b.batches = append(b.batches, singularBatchToElement(singularBatch))
+	b.l1OriginCheck = singularBatch.EpochHash.Bytes()[:20]
+	if len(b.batches) == 1 {
```
Can we reorder this, to check update when `if len(b.batches) == 0`, before append? It was bit unintuitive for me.

pcw109550 commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,533 @@
+	for _, batch := range b.batches {
+		blockTxCount := uint64(len(batch.Transactions))
+		blockTxCounts = append(blockTxCounts, blockTxCount)
+		blockTimstamps = append(blockTimstamps, batch.Timestamp)
```
`blockTimstamps`, `blockOriginNums` are unused.

pcw109550 commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,533 @@
+
+// decodePrefix parses data into b.spanBatchPrefix
+func (b *RawSpanBatch) decodePrefix(r *bytes.Reader) error {
+	relTimestamp, err := binary.ReadUvarint(r)
```
We must break down each field decoding to separate methods.

pcw109550 commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,533 @@
+
+// decodePayload parses data into b.spanBatchPayload
+func (b *RawSpanBatch) decodePayload(r *bytes.Reader) error {
+	blockCount, err := binary.ReadUvarint(r)
```
We must break down each field decoding to separate methods.

ImTei committed:
Adding error context

pcw109550 commented on op-node/rollup/derive/span_batch_txs.go:
```diff
@@ -0,0 +1,497 @@
+	return txs, nil
+}
+
+func (btx *spanBatchTxs) appendTx(txBytes []byte) error {
```
This method is unused.

pcw109550 commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -59,7 +62,20 @@ func (bq *BatchQueue) Origin() eth.L1BlockRef {
+	if len(bq.nextSpan) > 0 {
+		// If there are cached singular batches, pop first one and return.
+		nextBatch := bq.nextSpan[0]
+		bq.nextSpan = bq.nextSpan[1:]
```
Removing first element from slice is O(N). Maybe we can maintain an index? Or maybe this optimization is not worth it.

pcw109550 commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -111,7 +127,25 @@ func (bq *BatchQueue) NextBatch(ctx context.Context, safeL2Head eth.L2BlockRef)
+	}
+	spanBatch, ok := batch.(*SpanBatch)
+	if ok {
+		// If next batch is SpanBatch, converts it to SingularBatches.
```
```suggestion
		// If next batch is SpanBatch, convert it to SingularBatches.
```

pcw109550 commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -111,7 +127,25 @@ func (bq *BatchQueue) NextBatch(ctx context.Context, safeL2Head eth.L2BlockRef)
+		nextBatch.ParentHash = safeL2Head.Hash
+		return nextBatch, nil
+	}
+	return nil, nil
```
When code reaches here, we must raise an error.

pcw109550 commented on op-node/rollup/derive/batches.go:
```diff
@@ -45,36 +63,36 @@ func CheckBatch(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l
 
 	nextTimestamp := l2SafeHead.Time + cfg.BlockTime
-	if batch.Batch.Timestamp > nextTimestamp {
+	if batch.Timestamp > nextTimestamp {
```
Why use direct field instead of using `GetTimestamp` wrapper? Applies to other fields of singular batches.

pcw109550 commented on op-node/rollup/derive/batches.go:
```diff
@@ -29,13 +29,31 @@ const (
+			log.Warn("received SingularBatch after SpanBatch hard fork")
+			return BatchDrop
+		}
+		singularBatch, _ := batch.Batch.(*SingularBatch)
```
Is it ok to not checking that casting is properly done?

pcw109550 commented on op-node/rollup/derive/batches.go:
```diff
@@ -141,3 +159,130 @@ func CheckBatch(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l
+				log.Warn("batch is for different L1 chain, epoch hash does not match", "expected", l1Block.Hash)
+				return BatchDrop
+			}
+			originChecked = true
```
We can break here.

pcw109550 commented on op-node/rollup/derive/batches.go:
```diff
@@ -141,3 +159,130 @@ func CheckBatch(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l
+				// then allow the batch if it was the right thing to do to maintain the L2 time >= L1 time invariant.
+				// We only check batches that do not advance the epoch, to ensure epoch advancement regardless of time drift is allowed.
+				if !originAdvanced {
+					if originIdx+1 >= len(l1Blocks) {
```
```suggestion
					if originIdx >= len(l1Blocks) {
```

pcw109550 commented on op-node/rollup/derive/batches_test.go:
```diff
@@ -526,3 +541,661 @@ func TestValidBatch(t *testing.T) {
+			Expected:         BatchDrop,
+		},
+		{
+			Name:             "epoch hash wrong - long span",
```
Memo: newly added test for span batch.

pcw109550 commented on op-node/rollup/derive/batches_test.go:
```diff
@@ -526,3 +541,661 @@ func TestValidBatch(t *testing.T) {
+			Expected:         BatchDrop,
+		},
+		{
+			Name:             "sequencer time drift on same epoch with non-empty txs - long span",
```
Memo: newly added test for span batch.

pcw109550 commented on op-node/rollup/derive/batches_test.go:
```diff
@@ -526,3 +541,661 @@ func TestValidBatch(t *testing.T) {
+			Expected:         BatchUndecided, // we have to wait till the next epoch is in sight to check the time
+		},
+		{
+			Name:             "sequencer time drift on same epoch with empty txs and no next epoch in sight yet - long span",
```
Memo: newly added test for span batch.

pcw109550 commented on op-node/rollup/derive/batches_test.go:
```diff
@@ -526,3 +541,661 @@ func TestValidBatch(t *testing.T) {
+			Expected:         BatchDrop, // dropped because it could have advanced the epoch to B
+		},
+		{
+			Name:             "sequencer time drift on same epoch with empty txs and but in-sight epoch that invalidates it - long span",
```
Memo: newly added test for span batch.

pcw109550 commented on op-node/rollup/derive/batches_test.go:
```diff
@@ -526,3 +541,661 @@ func TestValidBatch(t *testing.T) {
+			Expected:         BatchDrop,
+		},
+		{
+			Name:             "batch with L2 time before L1 time - long span",
```
Memo: newly added test for span batch.

pcw109550 commented on op-node/rollup/derive/batches_test.go:
```diff
@@ -16,25 +18,38 @@ import (
+	L2SafeHead       eth.L2BlockRef
+	OriginChangedBit uint
+	SingularBatches  []*SingularBatch
+	L1InclusionBlock eth.L1BlockRef
```
Why maintain L1InclusionBlock, SingularBatches, TxData, OriginChangedBit instead of single span batch(type `Batch`)? Is it because too nested?

pcw109550 commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -148,19 +202,33 @@ func TestBatchQueueEager(t *testing.T) {
-	errors := []error{nil, nil, nil, nil, nil, nil, nil, io.EOF}
+	singularBatches := []*SingularBatch{b(cfg.L2ChainID, 12, l1[0]), b(cfg.L2ChainID, 14, l1[0]), b(cfg.L2ChainID, 16, l1[0]), b(cfg.L2ChainID, 18, l1[0]), b(cfg.L2ChainID, 20, l1[0]), b(cfg.L2ChainID, 22, l1[0]), nil}
+	errors := []error{nil, nil, nil, nil, nil, nil, io.EOF}
+	inputErrors := errors
```
Why copy `inputErrors`? Lets start with `inputErrors`.

pcw109550 commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -169,12 +237,13 @@ func TestBatchQueueEager(t *testing.T) {
+		} else {
+			require.Equal(t, singularBatches[i], b)
 			safeHead.Number += 1
 			safeHead.Time += 2
```
We may use `cfg.BlockTime` instead of `2`.

pcw109550 commented on op-node/rollup/derive/channel_out.go:
```diff
@@ -99,40 +125,119 @@ func (co *ChannelOut) AddBlock(block *types.Block) (uint64, error) {
+}
+
+// writeSpanBatch appends a SingularBatch to the channel's SpanBatch.
+// A channel can have only one SpanBatch. And compressed results should not be accessible until the channel is closed, since the prefix can be changed.
```
```suggestion
// A channel can have only one SpanBatch. And compressed results should not be accessible until the channel is closed, since the prefix or payload can be changed.
```

pcw109550 commented on op-node/rollup/derive/channel_out.go:
```diff
@@ -145,24 +250,36 @@ func (co *ChannelOut) InputBytes() int {
+	if err := co.compress.Flush(); err != nil {
+		return err
+	}
+	if co.batchType == SpanBatchType && co.closed && co.reader.Len() == 0 && co.compress.Len() > 0 {
```
```suggestion
	if co.batchType == SpanBatchType && co.closed && co.ReadyBytes() == 0 && co.compress.Len() > 0 {
```

Also can we apply `co.ReadyBytes()`s to other implementations?

pcw109550 commented on op-node/rollup/derive/channel_out.go:
```diff
@@ -99,40 +125,119 @@ func (co *ChannelOut) AddBlock(block *types.Block) (uint64, error) {
+		}
+		_, err := io.Copy(co.reader, co.compress)
+		if err != nil {
+			return 0, err
```
I think we should return `written` to follow the convention in `writeSingularBatch`. If we write to `reader`, its size must be returned.

ImTei replied:
Use Write()

pcw109550 commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -127,12 +134,14 @@ type channelBuilder struct {
 	}
-	co, err := derive.NewChannelOut(c)
+	// Creates a spanBatchBuilder contains information requires to build SpanBatch
+	spanBatchBuilder := derive.NewSpanBatchBuilder(cfg.ParentRef.L1Origin.Number, rcfg.Genesis.L2Time, rcfg.L2ChainID)
```
We do not have to always initialize `spanBatchBuilder`. if singular batch, lets set `spanBatchBuilder` as nil.

ImTei committed:
Apply small fixes from code review

ImTei committed:
Move spanBatchSIgnature to txs file

ImTei committed:
Handling batch type assertion failure

pcw109550 committed:
Add skeleton

pcw109550 committed:
Merge branch 'tei/span-batch-research' into pcw109550/span-batch-type-unit-test

ImTei committed:
Fix span_batch_tx.go

ImTei committed:
Reset ChannelOut reader when flush failed

ImTei committed:
TestValidSpanBatch refactoring

ImTei committed:
Revert unrelated test in channel_builder_test.go

ImTei committed:
Revert unrelated test in channel_test.go

ImTei committed:
Run ChannelManagerCloseAllTxsFailed

ImTei committed:
Revert unrelated test in channel_builder_test.go

ImTei committed:
Use SafeL2 to check empty syncStatus

ImTei committed:
Set SpanBatchTime into e2e system test config

ImTei committed:
Add batch queue test cases

ImTei committed:
Remove ChannelOut.fullErr

ImTei committed:
Add hard fork test in batches_test.go

ImTei committed:
Add channel manager hard fork test

ImTei committed:
Fix TestReorgFlipFlop for span batch

Fix l2_batcher reorg

ImTei committed:
Add an e2e test: TestDropSpanBatchBeforeHardfork

pcw109550 committed:
Add span batch txs unit test

pcw109550 committed:
Merge branch 'tei/span-batch-research' into pcw109550/span-batch-type-unit-test

ImTei committed:
Update span batch spec to allow overlapping batch

ImTei committed:
Add e2e test case to test overlap batch after reorg

ImTei committed:
Allow singular batch after hard fork

ImTei committed:
Add BatchType flag to op-batcher

Remove auto batch type detection feature

ImTei committed:
Apply code review

ImTei committed:
Update variable name

pcw109550 committed:
Add span batch tx unit test

pcw109550 committed:
Add more span batch txs unit test

pcw109550 committed:
Fix test name

ImTei committed:
Fetch safe blocks directly in CheckBatch

pcw109550 committed:
Refactor span batch prefix encode/decode

ImTei committed:
Add L2BlockRefByNumber method to OracleEngine

pcw109550 committed:
Refactor span batch payload encode/decode

ImTei committed:
Use new CheckBatch() in batch queue

Add unit test cases

ImTei committed:
Fix MockL2Client to return expected error

pcw109550 committed:
Handle empty list

pcw109550 committed:
Refactor

pcw109550 committed:
Add span batch unit test and refactor

pcw109550 committed:
Rename for clarity

pcw109550 committed:
Add batch unit test

pcw109550 committed:
Remove newline

pcw109550 committed:
Add batch fuzz tests

pcw109550 committed:
Merge branch 'tei/span-batch-research' into pcw109550/span-batch-type-unit-test

pcw109550 committed:
Add batch interface method tests

pcw109550 committed:
Avoid oom while decoding

pcw109550 committed:
Avoid oom while decoding

pcw109550 committed:
Avoid oom while decoding

ImTei committed:
Fix batch queueing

ImTei committed:
Fix processing overlapping batch

- Compare overlapping blocks' transactions excluding deposit txs
- Fix finding L1 origin blocks from the given L1blocks in Checkbatch()
- GetSingularBatches returns excluding overlapped blocks

ImTei committed:
Add missing line

ImTei committed:
Update span-batches.md

pcw109550 committed:
Set error when span batch field size too big

ImTei committed:
Add old epoch check

pcw109550 committed:
Proper error handling

pcw109550 committed:
Add max field length tests

pcw109550 committed:
Rename

pcw109550 committed:
Remove implicit usage of chainID

pcw109550 committed:
Styles and comments

pcw109550 committed:
Remove redundant comment

ImTei committed:
Small fixes from code review

ImTei committed:
Fix typo

ImTei committed:
Merge pull request #24 from testinprod-io/tei/span-batch-overlap-impl

Overlapping span batch implementation

ImTei committed:
Merge pull request #23 from testinprod-io/tei/span-batch-overlap

Allow overlapping span batch

pcw109550 committed:
Merge branch 'tei/span-batch-research' into pcw109550/span-batch-type-unit-test

pcw109550 committed:
Apply updated function signature

pcw109550 committed:
Remove redundancy

pcw109550 committed:
Remove TODOs

ImTei committed:
advance batch queue origin

ImTei committed:
Merge branch 'develop' into tei/span-batch-research

ImTei committed:
Rename span batch methods

ImTei committed:
Remove closing channel when hard fork activated

ImTei committed:
Devnet configs for span batch

ImTei committed:
Add TestAcceptSingularBatchAfterHardfork e2e test

ImTei committed:
Add e2e tests for span batch

ImTei committed:
Fix arg name

ImTei committed:
reorder packages

pcw109550 committed:
Refactor span batch encode/decode tx

pcw109550 committed:
Refactor

pcw109550 committed:
Support tests for span batch tx encapsulation

pcw109550 committed:
Add comments for bitlists

pcw109550 committed:
Capitalize

ImTei committed:
Merge pull request #22 from testinprod-io/pcw109550/span-batch-type-unit-test

Span Batch Type Unit Tests

ImTei committed:
Apply method name change

pcw109550 committed:
Trigger ci

pcw109550 committed:
Revert for ci

pcw109550 committed:
Remove unused method

pcw109550 committed:
Make linter happy

ImTei committed:
Remove maxTs from tests

pcw109550 committed:
Fix fuzz test to ignore span batch

ImTei committed:
Merge branch 'tei/span-batch-research' of github.com:testinprod-io/optimism into tei/span-batch-research

ImTei committed:
Handle fetchSyncStatus failure in batcher main loop

ImTei committed:
Fix lint errors

ImTei committed:
Fix markdown lint error

pcw109550 commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -127,12 +134,12 @@ type channelBuilder struct {
 // newChannelBuilder creates a new channel builder or returns an error if the
 // channel out could not be created.
-func newChannelBuilder(cfg ChannelConfig) (*channelBuilder, error) {
+func newChannelBuilder(cfg ChannelConfig, spanBatchBuilder *derive.SpanBatchBuilder) (*channelBuilder, error) {
```
Why add parameter `spanBatchBuilder`? `cfg` already has `BatchType` so `newChannelBuilder` method can initialize itself. 
-> No, newChannelBuilder does not have enough info to initialize spanBatchBuilder.

pcw109550 replied:
Hmm then how about adding all the necessary fields to `ChannelConfig` struct, to initialize spanbatchbuilder when necessary?

pcw109550 commented on op-batcher/batcher/channel_builder_test.go:
```diff
@@ -143,10 +153,12 @@ func newMiniL2BlockWithNumberParent(numTx int, number *big.Int, parent common.Ha
 	txs = append(txs, types.NewTx(l1InfoTx))
 	for i := 0; i < numTx; i++ {
-		txs = append(txs, types.NewTx(&types.DynamicFeeTx{}))
+		txs = append(txs, types.NewTx(l1InfoTx), testutils.RandomTx(rng, new(big.Int).SetUint64(rng.Uint64()), signer))
```
Must update comments:
> // If numTx > 0, that many empty DynamicFeeTxs will be added to the txs.

Now random tx is added.

pcw109550 replied:
Method name may be changed because the block is now not `mini`.

pcw109550 replied:
Using `testutils.RandomL2Block`seems to be more appropriate.

pcw109550 commented on op-batcher/batcher/channel_builder_test.go:
```diff
@@ -143,10 +153,12 @@ func newMiniL2BlockWithNumberParent(numTx int, number *big.Int, parent common.Ha
 		panic(err)
 	}
 
+	rng := rand.New(rand.NewSource(123))
```
Initializing `rand.Rand` inside callee is an anti pattern. Controlling the randomness becomes implicit.

pcw109550 commented on op-batcher/batcher/channel_builder_test.go:
```diff
@@ -480,37 +512,92 @@ func TestChannelBuilder_OutputFramesWorks(t *testing.T) {
+	require.NoError(t, cb.OutputFrames())
+
+	// There should be many frames in the channel builder now
+	require.Greater(t, cb.PendingFrames(), 1)
```
Can `cb.PendingFrames() == 1`?

ImTei replied:
yes

pcw109550 commented on op-batcher/batcher/channel_builder_test.go:
```diff
@@ -480,37 +512,92 @@ func TestChannelBuilder_OutputFramesWorks(t *testing.T) {
+	for i := 0; i < cb.numFrames-1; i++ {
+		require.Len(t, cb.frames[i].data, int(channelConfig.MaxFrameSize))
+
+	}
+}
```
Lets check last frame size too.
```suggestion
         require.LessOrEqual(t, len(cb.frames[len(cb.frames) - 1].data), int(channelConfig.MaxFrameSize))
	}
```

pcw109550 commented on op-batcher/batcher/channel_builder_test.go:
```diff
@@ -480,37 +512,92 @@ func TestChannelBuilder_OutputFramesWorks(t *testing.T) {
+	require.Greater(t, cb.PendingFrames(), 1)
+	for i := 0; i < cb.numFrames-1; i++ {
+		require.Len(t, cb.frames[i].data, int(channelConfig.MaxFrameSize))
+
```
Remove newline.

ImTei committed:
Add go-e2e-span-batch-tests pipeline

pcw109550 commented on op-batcher/batcher/channel_builder_test.go:
```diff
@@ -521,46 +608,50 @@ func TestChannelBuilder_OutputFramesMaxFrameIndex(t *testing.T) {
 			Number:     common.Big0,
 		}, nil, nil, nil, trie.NewStackTrie(nil))
 		l1InfoTx, _ := derive.L1InfoDeposit(0, eth.BlockToInfo(lBlock), eth.SystemConfig{}, false)
-		txs := []*types.Transaction{types.NewTx(l1InfoTx)}
```
We can use `testutils.RandomL2Block`.

pcw109550 commented on op-program/client/l2/engine.go:
```diff
@@ -105,6 +105,18 @@ func (o *OracleEngine) L2BlockRefByHash(ctx context.Context, l2Hash common.Hash)
+	if hash == (common.Hash{}) {
+		return eth.L2BlockRef{}, ErrNotFound
+	}
+	block := o.backend.GetBlockByHash(hash)
+	if block == nil {
+		return eth.L2BlockRef{}, ErrNotFound
+	}
+	return derive.L2BlockToBlockRef(block, &o.rollupCfg.Genesis)
```
```suggestion
	return derive.L2BlockRefByHash(ctx, hash)
```

pcw109550 commented on op-node/rollup/derive/channel_out_test.go:
```diff
@@ -28,8 +28,12 @@ func (s *nonCompressor) FullErr() error {
 	return nil
 }
 
+func (s *nonCompressor) ForceWrite(p []byte) (int, error) {
+	return 0, nil
+}
+
```
`ForceWrite` never used.

```suggestion
```

pcw109550 commented on op-batcher/batcher/channel_builder_test.go:
```diff
@@ -601,9 +693,11 @@ func TestChannelBuilder_Reset(t *testing.T) {
 	require.Equal(t, timeout, cb.timeout)
 	require.NoError(t, cb.fullErr)
+	if batchType == derive.SingularBatchType {
+		require.Greater(t, len(cb.frames), 1)
```
May I ask why we check this only on singularbatch?

pcw109550 commented on op-batcher/batcher/channel_manager_test.go:
```diff
@@ -153,7 +192,7 @@ func TestChannelManager_Clear(t *testing.T) {
 	require.Empty(m.txChannels)
```
We need to check `m.lastProcessedBlock` is also set.
How about
```go
m.Clear(nil)
...
require.Nil(m.lastProcessedBlock)
```
?

ImTei committed:
decrease test size

pcw109550 commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -45,18 +50,21 @@ type channelManager struct {
 	}
 }
 
 // Clear clears the entire state of the channel manager.
```
Now entire state except `lastProcessedBlock`. Must update comments.

pcw109550 commented on op-batcher/batcher/driver.go:
```diff
@@ -136,7 +145,11 @@ func (l *BatchSubmitter) Start() error {
-	l.state.Clear()
+	syncStatus, err := fetchSyncStatus(l.shutdownCtx, l.RollupNode, l.NetworkTimeout)
+	if err != nil {
+		return nil
```
Is this intended? `err` is not nil but return nil?

pcw109550 commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -45,18 +50,21 @@ type channelManager struct {
 }
 
 // Clear clears the entire state of the channel manager.
 // It is intended to be used after an L2 reorg.
```
Actually we are not always using when L2 reorg. We use while startup etc, so we may change the comment.

pcw109550 commented on op-e2e/actions/reorg_test.go:
```diff
@@ -189,8 +189,15 @@ func TestReorgFlipFlop(gt *testing.T) {
+		require.Equal(t, verifier.L2Safe(), verifier.L2Unsafe(), "head is at safe block after L1 reorg")
+	} else {
+		// after span batch hard fork
+		require.Zero(t, verifier.L2Safe().Number, "safe head is at genesis block because partially valid span batch not accepted")
```
Please clarify why partially valid span batch is not accepted. I think we should not use the term `partially valid`. Instead, use wrong l1 origin because of l1 reorg.

ImTei committed:
Fix batcher unit tests

pcw109550 commented on op-e2e/actions/reorg_test.go:
```diff
@@ -762,3 +770,108 @@ func TestConflictingL2Blocks(gt *testing.T) {
+		sequencer.ActL2PipelineFull(t)
+		sequencer.ActL2StartBlock(t)
+		if sequencer.derivation.UnsafeL2Head().Number == 11 {
+			// include a user tx at L2 block #12
```
Please add the reason why we add tx in block 12.
Reason: Perform state transition by inserting tx.

ImTei committed:
Apply code review related tests

ImTei committed:
Update L2 oracle engine

ImTei committed:
apply code review

pcw109550 commented on op-e2e/actions/l2_batcher.go:
```diff
@@ -103,31 +106,41 @@ func (s *L2Batcher) Buffer(t Testing) error {
+		s.l2SubmittedBlock = syncStatus.SafeL2
+		s.l2BufferedBlock = syncStatus.SafeL2
+		s.l2ChannelOut = nil
+		block, err = s.l2.BlockByNumber(t.Ctx(), big.NewInt(int64(s.l2BufferedBlock.Number+1)))
+		require.NoError(t, err, "need l2 block %d from sync status", s.l2SubmittedBlock.Number+1)
```
We do not need this.
```suggestion
```

pcw109550 commented on op-e2e/actions/l2_batcher.go:
```diff
@@ -140,23 +153,23 @@ func (s *L2Batcher) Buffer(t Testing) error {
 			})
 			require.NoError(t, e, "failed to create compressor")
-			ch, err = derive.NewChannelOut(c)
+			batchType := derive.SingularBatchType
+			var spanBatchBuilder *derive.SpanBatchBuilder = nil
+			if s.rollupCfg.IsSpanBatch(block.Time()) {
+				batchType = derive.SpanBatchType
+				spanBatchBuilder = derive.NewSpanBatchBuilder(s.l2BufferedBlock.L1Origin.Number, s.rollupCfg.Genesis.L2Time, s.rollupCfg.L2ChainID)
+			}
+			ch, err = derive.NewChannelOut(c, uint(batchType), spanBatchBuilder)
```
Minor suggestion
```suggestion
			var batchType uint := derive.SingularBatchType
			var spanBatchBuilder *derive.SpanBatchBuilder = nil
			if s.rollupCfg.IsSpanBatch(block.Time()) {
				batchType = derive.SpanBatchType
				spanBatchBuilder = derive.NewSpanBatchBuilder(s.l2BufferedBlock.L1Origin.Number, s.rollupCfg.Genesis.L2Time, s.rollupCfg.L2ChainID)
			}
			ch, err = derive.NewChannelOut(c, batchType, spanBatchBuilder)
```

ImTei committed:
Fix batcher lint error

ImTei committed:
Update l2_batcher.go

ImTei committed:
Log fetchSyncStatus error in batcher driver

pcw109550 commented on op-e2e/actions/span_batch_test.go:
```diff
@@ -0,0 +1,342 @@
+	for i := int64(1); i < int64(verifier.L2Safe().Number); i++ {
+		block, _ := verifCl.BlockByNumber(t.Ctx(), big.NewInt(i))
+		require.NoError(t, err)
+		// because verifier drops every span batch, it should generate empty batches.
```
```suggestion
		// because verifier drops every span batch, it should generate empty blocks.
```

pcw109550 commented on op-e2e/setup.go:
```diff
@@ -658,6 +661,11 @@ func (cfg SystemConfig) Start(t *testing.T, _opts ...SystemConfigOption) (*Syste
 		return nil, fmt.Errorf("unable to start l2 output submitter: %w", err)
 	}
 
+	batchType := derive.SingularBatchType
```
```suggestion
	var batchType uint = derive.SingularBatchType
```

And drop explicit casting below.

ImTei committed:
apply code review

pcw109550 commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -203,27 +322,54 @@ func TestBatchQueueInvalidInternalAdvance(t *testing.T) {
+		} else {
+			require.Equal(t, expectedOutputBatches[i], b)
 			safeHead.Number += 1
 			safeHead.Time += 2
```
```suggestion
			safeHead.Time += cfg.BlockTime
```

pcw109550 commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -294,30 +441,49 @@ func TestBatchQueueMissing(t *testing.T) {
 	}
 
-	// The batches at 18 and 20 are skipped to stop 22 from being eagerly processed.
+	// The intputBatches at 18 and 20 are skipped to stop 22 from being eagerly processed.
```
```suggestion
	// The inputBatches at 18 and 20 are skipped to stop 22 from being eagerly processed.
```

pcw109550 commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -294,30 +441,49 @@ func TestBatchQueueMissing(t *testing.T) {
+	// errors will be returned by fakeBatchQueueInput.NextBatch()
+	inputErrors := []error{nil, nil}
+	// batches will be returned by fakeBatchQueueInput
+	var intputBatches []Batch
```
```suggestion
	var inputBatches []Batch
```

pcw109550 commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -294,30 +441,49 @@ func TestBatchQueueMissing(t *testing.T) {
 	input := &fakeBatchQueueInput{
-		batches: batches,
-		errors:  errors,
+		batches: intputBatches,
```
```suggestion
		batches: inputBatches,
```

ImTei committed:
Fix typo

pcw109550 commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -45,16 +49,74 @@ func mockHash(time uint64, layer uint8) common.Hash {
 		EpochHash:    epoch.Hash,
-		Transactions: []hexutil.Bytes{data},
-	}}
+		Transactions: []hexutil.Bytes{txData},
```
There will be always single tx per singularbatch. Is this intended?
